File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();
        int low = 0;
        int high = m-1;
        while(low<=high){
            int mid = (low+high)/2;
            if(matrix[mid][0]==target) return true;
            if(target<matrix[mid][0]){
                high = mid-1;
            }else{
                low = mid+1;
            }
        }
        if(high<0 || high>m-1) return false;
        int row = high;
        low=0;
        high = n-1;
        while(low<=high){
            int mid = (low+high)/2;
            if(matrix[row][mid]==target) return true;
            if(target<matrix[row][mid]){
                high = mid-1;
            }else{
                low = mid+1;
            }
        }
        return false;
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def binSearch(self,arr,low,high,target):
        while(low<=high):
            mid = (low+high)//2
            if(arr[mid]==target):
                return True
            elif(arr[mid]<target):
                low = mid+1
            else:
                high = mid-1
        return False
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0])
        rl = 0
        rh = m-1
        while(rl<=rh):
            rm = (rl+rh)//2
            if(matrix[rm][n-1]<target):
                rl = rm+1
            else:
                rh = rm-1
         #rl will have my required row
        if(rl<0 or rl>=m):
            return False
        if(self.binSearch(matrix[rl],0,n-1,target)):
            return True
        return False 
        